File: Leetcoding-Actions/Explore-Monthly-Challenges/2020-12/06-populatingNextRightPointersOfEachNode_ii.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val: int = 0, left: 'Node' = None, right: 'Node' = None, next: 'Node' = None):
        self.val = val
        self.left = left
        self.right = right
        self.next = next
"""

class Solution:
    def connect(self, root: 'Node') -> 'Node':
        '''
        T: O(n) and S: O(1) ignoring implicit q space
        '''
        if not root: return root
        q = deque()
        q.append(root)
        
        while q:
            size = len(q)
            count = 0
            
            prevNode = None
            
            while count < size:
                currNode = q.popleft()
                
                if not prevNode:
                    prevNode = currNode
                else:
                    prevNode.next = currNode
                    prevNode = currNode
                
                if currNode.left: q.append(currNode.left)
                if currNode.right: q.append(currNode.right)
                count += 1
                
        return root

File: Leetcoding-Actions/Explore-Monthly-Challenges/2020-12/06-populatingNextRightPointersOfEachNode_ii.java
/*
// Definition for a Node.
class Node {
    public int val;
    public Node left;
    public Node right;
    public Node next;

    public Node() {}
    
    public Node(int _val) {
        val = _val;
    }

    public Node(int _val, Node _left, Node _right, Node _next) {
        val = _val;
        left = _left;
        right = _right;
        next = _next;
    }
};
*/

class Solution {
    public Node connect(Node root) {
        if (root == null) return root;
        
        ArrayDeque<Node> q = new ArrayDeque<>();
        q.addLast(root);
        
        while (!q.isEmpty()) {
            int size = q.size();
            int count = 0;
            Node prevNode = null;
            while (count < size) {
                Node currNode = q.removeFirst();
                
                if (prevNode == null) {
                    prevNode = currNode;
                } else {
                    prevNode.next = currNode;
                    prevNode = currNode;
                }
                
                if (currNode.left != null) {
                    q.addLast(currNode.left);
                }
                if (currNode.right != null) {
                    q.addLast(currNode.right);
                }
                count++;
            }
        }
        return root;
    }
}
